clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java

```diff
@@ -378,11 +378,17 @@ private void onErrorResponse(final R response, final long currentTimeMs) {
             case INVALID_REQUEST:
             case GROUP_MAX_SIZE_REACHED:
             case UNSUPPORTED_ASSIGNOR:
-            case UNSUPPORTED_VERSION:
                 logger.error("{} failed due to {}: {}", heartbeatRequestName(), error, errorMessage);
                 handleFatalFailure(error.exception(errorMessage));
                 break;
 
+            case UNSUPPORTED_VERSION:
+                message = "The cluster doesn't yet support the new consumer group protocol." +
+                        " Set group.protocol=classic to revert to the classic protocol until the cluster is upgraded.";
```

Comment:
If I'm not mistaken, there are multiple cases that could cause this error:

1. The cluster is < 4.0.0 and doesn't support the new `CONSUMER` group protocol
2. The cluster is >= 4.0.0 but isn't configured to use the new coordinator
3. The cluster is >= 4.0.0 and is configured to use the new coordinator but _isn't_ configured to support the new `CONSUMER` group protocol

The wording in the proposed error message doesn't distinguish the cases, but perhaps that's a good thing?

Reply:
Can we use the constants from `GroupProtocol` instead of hard-coding the strings here? Just helps when code spelunking.

Reply:
> If I'm not mistaken, there are multiple cases that could cause this error:
> 
> 1. The cluster is < 4.0.0 and doesn't support the new `CONSUMER` group protocol
> 2. The cluster is >= 4.0.0 but isn't configured to use the new coordinator
> 3. The cluster is >= 4.0.0 and is configured to use the new coordinator but _isn't_ configured to support the new `CONSUMER` group protocol
> 
> The wording in the proposed error message doesn't distinguish the cases, but perhaps that's a good thing?

After posting my review I noticed that @dajac [proposed the wording for the error message](https://github.com/apache/kafka/pull/17411#discussion_r1796499895). As such, it's already been discussed, so I'm fine with it as is. The message gives users what they need to resolve so 👍